log.go

```diff
@@ -39,89 +39,94 @@ import (
-	ltndLog = build.NewSubLogger("LTND", RootLogWriter.GenSubLogger)
-	peerLog = build.NewSubLogger("PEER", RootLogWriter.GenSubLogger)
-	rpcsLog = build.NewSubLogger("RPCS", RootLogWriter.GenSubLogger)
-	srvrLog = build.NewSubLogger("SRVR", RootLogWriter.GenSubLogger)
-	fndgLog = build.NewSubLogger("FNDG", RootLogWriter.GenSubLogger)
-	utxnLog = build.NewSubLogger("UTXN", RootLogWriter.GenSubLogger)
-	brarLog = build.NewSubLogger("BRAR", RootLogWriter.GenSubLogger)
-	atplLog = build.NewSubLogger("ATPL", RootLogWriter.GenSubLogger)
+	// by using the addSubLogger method. We declare all loggers with nil
+	// sub logger generators so we never run into a nil reference. But the
+	// SetupLoggers function should always be called to finish setting them
+	// up properly with a root logger.
+	ltndLog = build.NewSubLogger("LTND", nil)
+	peerLog = build.NewSubLogger("PEER", nil)
+	rpcsLog = build.NewSubLogger("RPCS", nil)
+	srvrLog = build.NewSubLogger("SRVR", nil)
+	fndgLog = build.NewSubLogger("FNDG", nil)
+	utxnLog = build.NewSubLogger("UTXN", nil)
+	brarLog = build.NewSubLogger("BRAR", nil)
+	atplLog = build.NewSubLogger("ATPL", nil)
 )
 
-// Initialize package-global logger variables.
-func init() {
-	setSubLogger("LTND", ltndLog, signal.UseLogger)
-	setSubLogger("ATPL", atplLog, autopilot.UseLogger)
-	setSubLogger("PEER", peerLog)
-	setSubLogger("RPCS", rpcsLog)
-	setSubLogger("SRVR", srvrLog)
-	setSubLogger("FNDG", fndgLog)
-	setSubLogger("UTXN", utxnLog)
-	setSubLogger("BRAR", brarLog)
-
-	addSubLogger("LNWL", lnwallet.UseLogger)
-	addSubLogger("DISC", discovery.UseLogger)
-	addSubLogger("NTFN", chainntnfs.UseLogger)
-	addSubLogger("CHDB", channeldb.UseLogger)
-	addSubLogger("HSWC", htlcswitch.UseLogger)
-	addSubLogger("CMGR", connmgr.UseLogger)
-	addSubLogger("BTCN", neutrino.UseLogger)
-	addSubLogger("CNCT", contractcourt.UseLogger)
-	addSubLogger("SPHX", sphinx.UseLogger)
-	addSubLogger("SWPR", sweep.UseLogger)
-	addSubLogger("SGNR", signrpc.UseLogger)
-	addSubLogger("WLKT", walletrpc.UseLogger)
-	addSubLogger("ARPC", autopilotrpc.UseLogger)
-	addSubLogger("INVC", invoices.UseLogger)
-	addSubLogger("NANN", netann.UseLogger)
-	addSubLogger("WTWR", watchtower.UseLogger)
-	addSubLogger("NTFR", chainrpc.UseLogger)
-	addSubLogger("IRPC", invoicesrpc.UseLogger)
-	addSubLogger("CHNF", channelnotifier.UseLogger)
-	addSubLogger("CHBU", chanbackup.UseLogger)
-	addSubLogger("PROM", monitoring.UseLogger)
-	addSubLogger("WTCL", wtclient.UseLogger)
-	addSubLogger("PRNF", peernotifier.UseLogger)
-	addSubLogger("CHFD", chanfunding.UseLogger)
-
-	addSubLogger(routing.Subsystem, routing.UseLogger, localchans.UseLogger)
-	addSubLogger(routerrpc.Subsystem, routerrpc.UseLogger)
-	addSubLogger(chanfitness.Subsystem, chanfitness.UseLogger)
-	addSubLogger(verrpc.Subsystem, verrpc.UseLogger)
+// SetupLoggers initializes all package-global logger variables.
+func SetupLoggers(root *build.RotatingLogWriter) {
+	ltndLog = build.NewSubLogger("LTND", root.GenSubLogger)
```

Comment:
Can we somehow combine this with the above decelerations? So people can add new "base package' loggers in _one_ place?

Reply:
Yes, makes sense. I'm not sure if what I did was what you had in mind but it seems to work.